src/layers/core/geojson-layer/geojson-layer.js

```diff
@@ -155,64 +120,116 @@ export default class GeoJsonLayer extends Layer {
-        onHover,
-        onClick
+    const {subLayers: {pointFeatures, lineFeatures, polygonFeatures}} = this.state;
+    const {id, getStrokeColor, getStrokeWidth, getFillColor, getHeight} = this.props;
+    const {extruded, wireframe} = this.props;
+
+    let {drawPoints, drawLines, drawPolygons, fillPolygons} = this.props;
+    drawPoints = drawPoints && pointFeatures && pointFeatures.length > 0;
+    drawLines = drawLines && lineFeatures && lineFeatures.length > 0;
+    drawPolygons = drawPolygons && polygonFeatures && polygonFeatures.length > 0;
+    fillPolygons = fillPolygons && polygonFeatures && polygonFeatures.length > 0;
+
+    // Filled Polygon Layer
+    const polygonFillLayer = fillPolygons && new PolygonLayer(Object.assign({}, this.props, {
+      id: `${id}-polygon-fill`,
+      data: polygonFeatures,
+      getPolygon: f => Container.get(f, 'geometry.coordinates'),
+      getHeight,
+      getColor: getFillColor,
+      extruded,
+      wireframe: false,
+      onHover: this._onHoverPolygon,
+      onClick: this._onClickPolygon
+    }));
+
+    // Polygon outline or wireframe
+    let polygonOutlineLayer = null;
+    if (drawPolygons && extruded && wireframe) {
+      polygonOutlineLayer = new PolygonLayer(Object.assign({}, this.props, {
+        id: `${id}-polygon-wireframe`,
+        data: polygonFeatures,
+        getPolygon: f => Container.get(f, 'geometry.coordinates'),
+        getHeight,
+        getColor: getFillColor,
+        extruded: true,
+        wireframe: true,
+        onHover: this._onHoverPolygon,
+        onClick: this._onClickPolygon
       }));
-
-    const pathLayer = pathLayerData &&
-      new PathLayer(Object.assign({}, this.props, {
-        id: `${id}-path-sublayer`,
-        data: pathLayerData,
-        getPaths: this._extractPaths,
+    } else if (drawPolygons) {
+      polygonOutlineLayer = new PathLayer(Object.assign({}, this.props, {
+        id: `${id}-polygon-outline`,
+        data: polygonFeatures,
+        getPaths: f => Container.get(f, 'geometry.coordinates'),
         getColor: getStrokeColor,
         getWidth: getStrokeWidth,
-        getOpacity: getStrokeOpacity,
-        pickable: false,
-        onHover,
-        onClick
-      }));
-
-    const polygonLayer = polygonLayerData &&
-      new PolygonLayer(Object.assign({}, this.props, {
-        id: `${id}-polygon-sublayer`,
-        data: polygonLayerData,
-        getPolygons: this.getFeaturePolygons,
-        getColor: getFillColor,
-        getOpacity: getFillOpacity,
-        pickable: false,
-        onHover,
-        onClick
+        onHover: this._onHoverPolygon,
+        onClick: this._onClickPolygon
       }));
+    }
 
-    return [polygonLayer, pathLayer, pointLayer].filter(Boolean);
+    const lineLayer = drawPolygons && drawLines && new PathLayer(Object.assign({}, this.props, {
+      id: `${id}-line-paths`,
+      data: lineFeatures,
+      getPaths: f => Container.get(f, 'geometry.coordinates'),
```

Comment:
Understood. A problem with heavily transforming the data in the GeoJson layer (apart from possible performance issues) is that it gets harder to track the connection to the original features, e.g. during picking as discussed above. I will clearly have to give this more thought.